src/audio_unit/looper.rs

```diff
@@ -88,24 +110,72 @@ impl Looper {
                 }
             }
             Playing { position, total } => {
+                self.process_samples_playing(position, total, input, output);
+            }
+            PlayingAwaitingOverdub { position, total } => {
+                self.process_samples_playing(position, total, input, output);
+            }
+            Overdubbing { position, total } => {
                 let next_position = position + output.len();
 
                 if next_position <= total {
                     output.copy_from_slice(&self.buffer[position..next_position]);
+
+                    for sample_number in 0..input.len() {
+                        self.buffer[position + sample_number] =
+                            self.buffer[position + sample_number] + input[sample_number];
+                    }
+
                     let position = if next_position == total {
                         0
                     } else {
                         next_position
                     };
 
-                    self.state = Playing { position, total }
+                    self.state = match next_position == total {
+                        true => {
+                            println!("looper: done overdubbing");
+                            Playing { position, total }
+                        }
+                        _ => Overdubbing { position, total },
+                    }
                 } else {
                     self.process_samples_wrap_around(position, total, input, output);
                 }
             }
         };
     }
 
+    fn process_samples_playing(
+        &mut self,
+        position: usize,
+        total: usize,
+        input: &[f32],
+        output: &mut [f32],
+    ) {
+        let next_position = position + output.len();
+
+        if next_position <= total {
+            output.copy_from_slice(&self.buffer[position..next_position]);
+            let position = if next_position == total {
+                0
+            } else {
+                next_position
+            };
+
+            self.state = match self.state {
+                PlayingAwaitingOverdub { .. } if next_position == total => {
```

Comment:
Could use this syntax above instead of `_` for the fields.

Reply:
I think this suggestion was trumped by switching those cases to just `_ => Off`, but that's useful to connect those dots conceptually